In [1]:
import re #importação do regex
# Leitura do arquivo csv
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', None) #Não limita a exibição dos n°s de registros
pd.options.display.max_columns = None #Não limita a exibição das colunas dos DataFrames
pd.set_option('display.max_colwidth', None) #Não  limita a exibição do conteúdo das células
df_bd = pd.read_csv("TabelaGeralBd.csv", low_memory=False)

In [2]:
print("Quantidades de linhas da tabela origem:",len(df_bd))

Quantidades de linhas da tabela origem: 35752


In [3]:
#Pego todos os registros que tem suporte igual a 'Fotografias'
df_bd_foto = df_bd[df_bd['suporte'] == 'Fotografias'] 
df_bd_foto

LIGAÇÃO  ALUNO (A)  TEMA: GRAU:  ORIENTADOR:  DEPTO:  ÁREA:  \
277       278        NaN    NaN   NaN          NaN     NaN    NaN   
278       279        NaN    NaN   NaN          NaN     NaN    NaN   
279       280        NaN    NaN   NaN          NaN     NaN    NaN   
280       281        NaN    NaN   NaN          NaN     NaN    NaN   
281       282        NaN    NaN   NaN          NaN     NaN    NaN   
282       283        NaN    NaN   NaN          NaN     NaN    NaN   
283       284        NaN    NaN   NaN          NaN     NaN    NaN   
284       285        NaN    NaN   NaN          NaN     NaN    NaN   
285       286        NaN    NaN   NaN          NaN     NaN    NaN   
286       287        NaN    NaN   NaN          NaN     NaN    NaN   
339       340        NaN    NaN   NaN          NaN     NaN    NaN   
351       352        NaN    NaN   NaN          NaN     NaN    NaN   
352       353        NaN    NaN   NaN          NaN     NaN    NaN   
609       610        NaN    NaN   NaN          NaN     NaN    NaN   
610       611        NaN    NaN   NaN          NaN     NaN    NaN   
636       637        NaN    NaN   NaN          NaN     NaN    NaN   
637       638        NaN    NaN   NaN          NaN     NaN    NaN   
665       666        NaN    NaN   NaN          NaN     NaN    NaN   
853       854        NaN    NaN   NaN          NaN     NaN    NaN   
854       855        NaN    NaN   NaN          NaN     NaN    NaN   
855       856        NaN    NaN   NaN          NaN     NaN    NaN   
856       857        NaN    NaN   NaN          NaN     NaN    NaN   
857       858        NaN    NaN   NaN          NaN     NaN    NaN   
1072     1073        NaN    NaN   NaN          NaN     NaN    NaN   
1073     1074        NaN    NaN   NaN          NaN     NaN    NaN   
1074     1075        NaN    NaN   NaN          NaN     NaN    NaN   
1075     1076        NaN    NaN   NaN          NaN     NaN    NaN   
1076     1077        NaN    NaN   NaN          NaN     NaN    NaN   
1077     1078        NaN    NaN   NaN          NaN     NaN    NaN   
1078     1079        NaN    NaN   NaN          NaN     NaN    NaN   
1079     1080        NaN    NaN   NaN          NaN     NaN    NaN   
1080     1081        NaN    NaN   NaN          NaN     NaN    NaN   
1081     1082        NaN    NaN   NaN          NaN     NaN    NaN   
1082     1083        NaN    NaN   NaN          NaN     NaN    NaN   
1083     1084        NaN    NaN   NaN          NaN     NaN    NaN   
1084     1085        NaN    NaN   NaN          NaN     NaN    NaN   
1085     1086        NaN    NaN   NaN          NaN     NaN    NaN   
1105     1106        NaN    NaN   NaN          NaN     NaN    NaN   
1106     1107        NaN    NaN   NaN          NaN     NaN    NaN   
1107     1108        NaN    NaN   NaN          NaN     NaN    NaN   
1187     1188        NaN    NaN   NaN          NaN     NaN    NaN   
1188     1189        NaN    NaN   NaN          NaN     NaN    NaN   
1189     1190        NaN    NaN   NaN          NaN     NaN    NaN   
1190     1191        NaN    NaN   NaN          NaN     NaN    NaN   
1191     1192        NaN    NaN   NaN          NaN     NaN    NaN   
1192     1193        NaN    NaN   NaN          NaN     NaN    NaN   
1193     1194        NaN    NaN   NaN          NaN     NaN    NaN   
1194     1195        NaN    NaN   NaN          NaN     NaN    NaN   
1195     1196        NaN    NaN   NaN          NaN     NaN    NaN   
1196     1197        NaN    NaN   NaN          NaN     NaN    NaN   
1197     1198        NaN    NaN   NaN          NaN     NaN    NaN   
1198     1199        NaN    NaN   NaN          NaN     NaN    NaN   
1199     1200        NaN    NaN   NaN          NaN     NaN    NaN   
1200     1201        NaN    NaN   NaN          NaN     NaN    NaN   
1201     1202        NaN    NaN   NaN          NaN     NaN    NaN   
1202     1203        NaN    NaN   NaN          NaN     NaN    NaN   
1203     1204        NaN    NaN   NaN          NaN     NaN    NaN   
1269

In [6]:
print("Quantidade de registros em fotografias", len(df_bd_foto))

Quantidade de registros em fotografias 209


In [7]:
#Verificação dos conteúdos de cada coluna
for col in df_bd_foto.columns.tolist(): #pega as colunas do DataFrame e transforma em list
    col_unique = df_bd_foto[col].unique()
    if(not(len(col_unique) == 1 and col_unique[0] != col_unique[0])): #verifica se tem algum valor nas colunas, isto é se a coluna não tem apenas um valor, e se este valor não é NaN
        print("================>",col, len(col_unique)) #Destaca com ====> o nome da coluna e o tamanho de seus registros
        print( col_unique[0:4]) #exibe os 4 primeiros valores da coluna

================> LIGAÇÃO 209
[278 279 280 281]
================> especificação 203
['Professor Emérito. Concessão do Título de Professor Emérito à Gilda de Mello e Souza  pela Congregação da Faculdade de Filosofia.'
 'Professor Emérito. Concessão do Título de Professor Emérito à Emília Viotti da Costa pela Congregação da Faculdade de Filosofia.'
 'Encontro/Congresso. Encontro (XII) de Professores Universitários Brasileiros de Literatura Portuguesa e Congresso (IV) Internacional de Estudos Pessoanos.'
 'Simpósio de Cultura Oriental e Cultura Ocidental. (Obs.: Fotos que faltam nºs. 18-25, 29, 31, 37).']
================> nreferencia 209
['P.33.29/NT 1303-1332' 'P.33.29/NT 1333-1335 - P. 34.40/NT 1336-1352'
 'P.10.6/NT 180-188; P.11.7/NT 189-218; P.12.8/NT 219-239'
 'P.23.19/NT 623-640; P.24.20/NT 641-700; P.25.21/NT 701-763']
================> numero 207
[129. 126.  16. 206.]
================> local 21
['São Paulo/Salão Nobre- FFLCH' 'Anfiteatro de Convenções/USP'
 'Departamento de Ciên

# Campos válidos de df_bd_foto 

    1. LIGAÇÃO 209
    2. especificação 203
    3. nreferencia 209
    4. numero 207
    5. local 21
    6. titulo 30
    7. data 21
    8. produção 4
    9. Sistema 54
    
## Campos possíveis de normalização
    1. local
    2. data
    3. produção
    9. Sistema

In [8]:
#Atualizo o  DataFrame apenas com as colunas que possuem valores
df_bd_foto = pd.DataFrame(
    {'nreferencia':df_bd_foto['nreferencia'], 
     'especificação':df_bd_foto['especificação'], 
     'numero':df_bd_foto['numero'],
     'ligação':df_bd_foto['LIGAÇÃO'],
     'sistema':df_bd_foto['Sistema'],
     'titulo':df_bd_foto['titulo'],
     'local':df_bd_foto['local'],
     'data':df_bd_foto['data'],
     'produção':df_bd_foto['produção']
    })


In [9]:
df_bd_foto

nreferencia  \
277                                         P.33.29/NT 1303-1332   
278                 P.33.29/NT 1333-1335 - P. 34.40/NT 1336-1352   
279      P.10.6/NT 180-188; P.11.7/NT 189-218; P.12.8/NT 219-239   
280   P.23.19/NT 623-640; P.24.20/NT 641-700; P.25.21/NT 701-763   
281                       P.27.23/NT 828-887; P.28.24/NT 888-926   
282                                          P.28.24//NT 940-963   
283                                               P.6.2/NT 45-50   
284                                               P.6.2/NT 51-53   
285                                               P.6.2/NT 54-83   
286                                         P.29.25/NT 1014-1016   
339                                               Cx.1;E.1/NT 56   
351                                                Cx.2;E.1/NT 1   
352                                                Cx.2;E.1/NT 2   
609                    P.31.27/NT 1205,1206 - 32.28/NT 1207-1235   
610                                         P.32.28/NT 1236-1272   
636                                                       118.20   
637                                                       118.21   
665                                               Cx.1;E.2/NT 73   
853                                               Cx.1;E.2/NT 72   
854                                               Cx.1;E.2/NT 71   
855                                               Cx.1;E.2/NT 70   
856                                               Cx.1;E.2/NT 69   
857                                               Cx.1;E.2/NT 68   
1072                                              Cx.2;E.2/NT 31   
1073                                              Cx.2;E.2/NT 32   
1074                                              Cx.2;E.2/NT 33   
1075                                              Cx.2;E.2/NT 34   
1076                                              Cx.2;E.2/NT 35   
1077                                              Cx.2;E.2/NT 36   
1078                                              Cx.2;E.2/NT 37   
1079                                              Cx.2;E.2/NT 38   
1080                                              Cx.2;E.2/NT 39   
1081                                              Cx.2;E.2/NT 40   
1082                                              Cx.2;E.2/NT 41   
1083                                              Cx.2;E.2/NT 42   
1084                                              Cx.2;E.2/NT 43   
1085                                              Cx.2;E.2/NT 44   
1105                                               Cx.2;E.1/NT 9   
1106                                              Cx.2;E.1/NT 10   
1107                                              Cx.2;E.1/NT 11   
1187                                      Cx.2; NT 119.9 -119.16   
1188                                      Cx.3; NT 119.17-119.24   
1189                                      Cx.4; NT 119.25-119.32   
1190                                      Cx.5; NT 119.33-119.37   
1191                                                       118.1   
1192                                                       118.2   
1193                                                       118.3   
1194                                                       118.4   
1195                                                       118.5   
1196                                                       118.6   
1197                                                       118.7   
1198                                                       118.8   
1199                                                       118.9   
1200                                                      118.10   
1201                                                      118.11   
1202                                                      118.12   
1203                                                      118.13   
1269                                              Cx.1;E.3/NT 80   
1270                                     

In [10]:
#verifico os tipos de valores do DF
df_bd_foto.dtypes

nreferencia       object
especificação     object
numero           float64
ligação            int64
sistema           object
titulo            object
local             object
data              object
produção          object
dtype: object

In [22]:
df_bd_foto.groupby('local')['local'].agg([np.size])

size
local                                                     
Anfiteatro de Convenções/USP                             2
Anfiteatro de História/USP                               2
Biblioteca do Depto.Geografia/USP                        1
Bienal do Livro/SP                                       2
CAPH-USP                                                33
Departamento de Ciências Sociais/USP                     1
Departamento de Geografia/USP                            2
Departamento de História/USP                             2
Departamento de Letras/USP                               1
IEB                                                      1
Prédio da Administração FFLCH/USP                        1
Prédio de Geografia e História                           2
Rio de Janeiro/RJ                                        1
Salão Nobre da FFLCH/USP                                 5
Salão Nobre da FFLCH/USP e Centro de Informática         1
São Paulo/Anfiteatro Geografia                           2
São Paulo/Salão Nobre- FFLCH                            18
São Paulo/Salão Nobre-FFLCH/Centro de Convenções-USP     1

In [21]:
df_bd_foto.local = df_bd_foto.local.replace('CAPH', 'CAPH-USP')
df_bd_foto.local = df_bd_foto.local.replace('São Paulo/Anfiteatro de Geografia', 'São Paulo/Anfiteatro Geografia')

In [28]:
df_bd_foto.groupby('data')['data'].agg([np.size])

size
data               
01-30/04/1986    25
05/07/1922        1
07/06/1911        1
1937/1979         1
1945              1
1973              1
1985              2
1986              4
1987              1
1988              2
1989              7
1990              7
1991              1
1992              5
1994              2
1996              1
1997              5
1998              5
1999              3

In [27]:
df_bd_foto.data = df_bd_foto.data.replace('s/d', None)

In [29]:
df_bd_foto.groupby('produção')['produção'].agg([np.size])

size
produção                 
CAPH                   41
Col/9 x 12,5 cm         1
Eduardo José Afonso    25

In [31]:
df_bd_foto.groupby('sistema')['sistema'].agg([np.size])

size
sistema                                            
B/P 18,0 x 24,0 cm                               67
B/P 22,0 x 30,0 cm                               25
BP                                                2
BP/10,5 x 23,5 cm                                 1
BP/11,0 x 24,0 cm                                 1
BP/12,0 x 18,0 cm                                 2
BP/12,5 x 23,0 cm                                 1
BP/14,5 x 22,5 cm                                 1
BP/14,5 x 23,5 cm                                 1
BP/15,0 x 23,0 cm                                 1
BP/15,0 x 24,0 cm                                 2
BP/15,5 x 17,5 cm                                 1
BP/15,5 x 22,5 cm                                 1
BP/15,5 x 23,0 cm                                 1
BP/15,5 x 23,5 cm                                 1
BP/16,0 x 22,5 cm                                 1
BP/16,0 x 23,0 cm                                 1
BP/16,0 x 23,5 cm                                 1
BP/16,5 x 22,5 cm                                 1
BP/16,5 x 23,0 cm                                 2
BP/17,0 x 17,5 cm                                 1
BP/17,0 x 19,0 cm                                 1
BP/17,0 x 19,5 cm                                 2
BP/17,0 x 22,0 cm                                 1
BP/17,0 x 22,5 cm                                 2
BP/17,0 x 23,0 cm                                 3
BP/17,5 x 17,5 cm                                 2
BP/17,5 x 18,0 cm                                 2
BP/17,5 x 18,5 cm                                 2
BP/17,5 x 19,5 cm                                 1
BP/17,5 x 20,5 cm                                 1
BP/17,5 x 21,0 cm                                 3
BP/17,5 x 21,5 cm                                 2
BP/17,5 x 22,5 cm                                 3
BP/17,5 x 23,0 cm                                 6
BP/17,5 x 23,5 cm                                 2
BP/17,5 x 24,0 cm                                 1
BP/18,0 x 19,0 cm                                 2
BP/18,0 x 22,0 cm                                 1
BP/18,0 x 23,0 cm                                 2
BP/18,0 x 23,5 cm                                 1
BP/18,0 x 24,0 cm                                 3
BP/29,5 x 39,0 cm                                 5
Col/10 x 15 cm                                    7
Col/10,0 x 15,0 cm                                8
Col/10x15 cm                                      2
Col/9 x 12,5 cm                                   8
Col/9 x 13 cm                                     1
Col/9,0 X 13,0 cm                                 1
Col/9,0 x 12,5 cm                                 7
Col/9,0 x 13,0 cm                                 9
PB e Col/24x30, 15x24, 15x20, 12x18, 17x24 cm     1
PB/26 x 20 cm                                     1

In [33]:
#vou trabalhar com a col nreferencia
s_referencia = df_bd_foto.nreferencia
s_referencia[s_referencia.duplicated()]#pego as referencias repetidas
#Não há, então esta tabela já está pronta pro merge

Series([], Name: nreferencia, dtype: object)

In [ ]:
df_bd_foto.to_csv("fotografia_geralBD.csv", index=False, encoding='UTF-8' )